### PR TITLE
Improve distributed language model training

### LINKED ISSUE #1933: Distributed language model gets worse results than training on a single GPU
**Describe the bug**
[I discovered](https://forums.fast.ai/t/should-training-distributed-with-bs-x-gpus-be-equivalent/42685/5) that training a language model using `to_distributed` on multiple GPUs gives worse results than with a single GPU.

I went ahead and tested this on another dataset to make sure it wasn't a quirk with my specific data and was able to replicate the problem on the `IMDB` dataset.

I've tracked down the problem to the way `torch.utils.data.distributed.DistributedSampler` is splitting up the data. I'm preparing a pull request that fixes this issue.

**Provide your installation details**
```text
=== Software === 
python        : 3.7.1
fastai        : 1.0.51
fastprogress  : 0.1.20
torch         : 1.0.0
nvidia driver : 410.66
torch cuda    : 10.0.130 / is available
torch cudnn   : 7401 / is enabled

=== Hardware === 
nvidia gpus   : 8
torch devices : 8
  - gpu0      : 8119MB | GeForce GTX 1070
  - gpu1      : 8119MB | GeForce GTX 1070
  - gpu2      : 8119MB | GeForce GTX 1070
  - gpu3      : 8119MB | GeForce GTX 1070
  - gpu4      : 8119MB | GeForce GTX 1070
  - gpu5      : 8119MB | GeForce GTX 1070
  - gpu6      : 8119MB | GeForce GTX 1070
  - gpu7      : 8119MB | GeForce GTX 1070

=== Environment === 
platform      : Linux-4.15.0-46-generic-x86_64-with-debian-stretch-sid
distro        : #49~16.04.1-Ubuntu SMP Tue Feb 12 17:45:24 UTC 2019
conda env     : Unknown
python        : /opt/conda/bin/python
sys.path      : /workspace/course-v3/nbs/dl1
/opt/conda/lib/python37.zip
/opt/conda/lib/python3.7
/opt/conda/lib/python3.7/lib-dynload

/opt/conda/lib/python3.7/site-packages
/opt/conda/lib/python3.7/site-packages/IPython/extensions
/root/.ipython
```

**To Reproduce**
I've [created a gist here](https://gist.github.com/yeldarby/07379fa373161282a6cf16b00c7f81d1) with the code I used for testing distributed language model training on the IMDB dataset.

**Expected behavior**
The baseline accuracy and loss from [`course-v3/lesson3-imdb.ipynb`](https://github.com/fastai/course-v3/blob/master/nbs/dl1/lesson3-imdb.ipynb) is as follows:

imdb-baseline

**Actual behavior**
Training on 8 GPUs using [the code from the gist](https://gist.github.com/yeldarby/07379fa373161282a6cf16b00c7f81d1) shows much worse results:

distributed-before

**Additional context**
I'm preparing a pull request with a proposed fix.

### PR DESCRIPTION
**Pull request by yeldarby**
This PR fixes #1933 by adding `DistributedChunkSampler` which is identical to `torch.utils.data.distributed.DistributedSampler` except that it doesn't permute the indices before distributing them.

I also added an option for users to pass a custom sampler when calling `to_distributed` if they want to override this functionality.

In my tests this has improved the performance of language models when training on multiple GPUs using `Distributed`.

Unfortunately, my multi-gpu instance was pre-empted a little bit before it finished 10 full epochs on the updated code. But the gains over `DistributedSampler` are clearly visible.

IMDB After
========
distributed-after

vs

IMDB Before
==========

distributed-before

The above screenshots are from a training run [with this code](https://gist.github.com/yeldarby/07379fa373161282a6cf16b00c7f81d1) on 8x GTX 1070. I saw similar results on a different language model on 8x 2080Ti.

### CODE REVIEW

**Comment by sgugger** [comment]
That's annoying that `DistributedSampler` doesn't have a shuffle option. I'll merge this but refine it a little bit later today to create a more general `DistributedSampler` that takes that option `shuffle=False` then we will use pass it the value of `train_dl.shuffle` so it will work in any situation we don't want the data to be randomly dispatched.

Thanks for digging into this!
